### PR TITLE
feat(start-campaign-button): warn before starting campaign without survey responses

### LINKED ISSUE #1578: Add warning upon starting a campaign without survey responses
**Is your feature request related to a problem? Please describe.**
Canned responses are sometimes mistakenly used in place of survey responses rather than as supplementary FAQ responses.(#984) Once texters have completed conversations using canned responses, it is tedious or impossible for admins to match those conversations back to survey data or tags.

**Describe the solution you'd like**
If an admin attempts to start a campaign that has no survey responses, a dialog box should warn that texters will not have the option to log survey data. This warning should link to documentation on [survey responses versus canned responses](https://docs.spokerewired.com/article/125-survey-responses-vs-canned-responses). The admin should be able to acknowledge and bypass the warning to start the campaign anyway, but the buttons should be configured to encourage them to go back and add surveys. 

> **Heads up! Campaign [ID] currently has no survey responses. Do you want to start this campaign anyway?**
⚫`Start this campaign` 🟢`Go back`

>Without survey responses, texters can only use [Tags](https://docs.spokerewired.com/article/67-tags) to log data on this campaign. Tags do not prepopulate scripts for texters and are not scoped to a single campaign.
Canned responses prepopulate scripts but log no data at all. ([Read more](https://docs.spokerewired.com/article/125-survey-responses-vs-canned-responses)) 
>For the best experience for you and your texters, we recommend adding survey responses before starting this campaign!

**Describe alternatives you've considered**
#1571 is the more passive first step of this request, adding a warning to the instructional copy within the canned responses section. This feature is requested in addition to the solution presented by #1571.

### PR DESCRIPTION
**Pull request by henryk1229**
## Description

This pr adds a `<Dialog />` component that warns an admin if they start a campaign with no survey responses, with a brief explanation of why they should add survey responses, as well as links to relevant docs. It prompts them to either confirm that they want to start the campaign, or cancel, so they can go back and add survey responses.

## Motivation and Context

It closes #1578 

## How Has This Been Tested?

Locally

## Screenshots (if appropriate):
Screenshot 2023-06-13 at 3 27 44 PM

https://github.com/politics-rewired/Spoke/assets/47220718/6aac0c52-a8b6-4d56-aac4-9d7b78a4afd2

https://github.com/politics-rewired/Spoke/assets/47220718/85d2e32a-386d-448a-9a72-a276c6f21bee

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have included updates for the documentation accordingly.

### CODE REVIEW

File: src/containers/AdminCampaignEdit/components/StartCampaignButton.tsx [hunk 1]
```diff
@@ -38,11 +47,32 @@ export const StartCampaignButton: React.FC<StartCampaignButtonProps> = (
 
   const tooltipText = requiresApproval ? "Superadmin approval required" : "";
 
+  const campaignInteractionSteps =
+    campaignStepsData?.campaign?.interactionSteps ?? [];
+
+  const campaignHasSurveyResponses = campaignInteractionSteps.length > 1;
```
Comment (henryk1229) [inline src/containers/AdminCampaignEdit/components/StartCampaignButton.tsx:53]: I couldn't tell from the ticket if we want/need to be more thorough than checking that a campaign has more than one interaction step to determine if the admin should be shown the warning dialog. Happy to add more logic + fields to the graphql query above, if there's additional info about the interaction steps needed.

Comment (ajohn25) [reply src/containers/AdminCampaignEdit/components/StartCampaignButton.tsx:53]: this is fine imo!

File: src/containers/AdminCampaignEdit/components/StartCampaignButton.tsx [hunk 2]
```diff
@@ -23,8 +25,15 @@ export const StartCampaignButton: React.FC<StartCampaignButtonProps> = (
   const { data, loading } = useGetCampaignStatusQuery({
     variables: { campaignId }
   });
+  const { data: campaignStepsData } = useGetCampaignInteractionStepsQuery({
+    variables: { campaignId }
+  });
   const [startCampaign] = useStartCampaignMutation();
 
+  // manage warning confirmation and confirmation dialog state
+  const [warningIsConfirmed, setWarningConfirmed] = useState<boolean>(false);
```
Comment (ajohn25) [inline src/containers/AdminCampaignEdit/components/StartCampaignButton.tsx:34]: Nit: We should create real conventions around this but I think this is more in line with what we usually do
```suggestion
  const [warningConfirmed, setWarningConfirmed] = useState<boolean>(false);
```

File: src/containers/AdminCampaignEdit/components/StartCampaignButton.tsx [hunk 3]
```diff
@@ -23,8 +25,15 @@ export const StartCampaignButton: React.FC<StartCampaignButtonProps> = (
...
+  const [showConfirmDialog, setShowConfirmDialog] = useState<boolean>(false);
```
Comment (ajohn25) [inline src/containers/AdminCampaignEdit/components/StartCampaignButton.tsx:35]: Nit (similar to above)
```suggestion
  const [confirmDialogOpen, setConfirmDialogOpen] = useState<boolean>(false);
```

File: src/containers/AdminCampaignEdit/components/StartCampaignButton.tsx [hunk 4]
```diff
@@ -53,19 +83,27 @@ export const StartCampaignButton: React.FC<StartCampaignButtonProps> = (
...
+          >
+            {startText}
+          </Button>
+        </span>
+      </Tooltip>
+      <CampaignSurveyResponseWarningDialog
+        open={showConfirmDialog}
+        campaignId={campaignId}
+        confirmWarningAndStartCampaign={confirmWarningAndStartCampaign}
+        setShowConfirmDialog={setShowConfirmDialog}
```
Comment (ajohn25) [inline src/containers/AdminCampaignEdit/components/StartCampaignButton.tsx:104]: Nit: Renaming these props in the child component follows a loose convention that I prefer for readability in both the parent and child as seen here
```suggestion
        onConfirm={confirmWarningAndStartCampaign}
        onClose={setShowConfirmDialog}
```

**Review by ajohn25 — APPROVED** [review]
Gotta love another dialog 🎉 My comments are all about unwritten conventions that I think we'll want 2nd thoughts from @hiemanshu about

File: src/containers/AdminCampaignEdit/components/StartCampaignButton.tsx (hunk 4 above)
Comment (hiemanshu) [reply src/containers/AdminCampaignEdit/components/StartCampaignButton.tsx:104]: And we also have the functions being `handleConfirmWarningAndStartCampaign`, so we know it's a handler.

**Review by hiemanshu — APPROVED** [review]
LGTM! I'd take care of the nits from Aashish!

File: src/containers/AdminCampaignEdit/components/StartCampaignButton.tsx [hunk 5]
```diff
@@ -38,11 +47,36 @@ export const StartCampaignButton: React.FC<StartCampaignButtonProps> = (
...
+
+  const handleConfirmWarningAndStartCampaign = useCallback(() => {
+    setWarningConfirmed(true);
+    setConfirmDialogOpen(false);
+    startCampaign({ variables: { campaignId } });
+  }, [startCampaign, campaignId, setWarningConfirmed, setConfirmDialogOpen]);
+
+  const handleCloseDialog = useCallback(() => {
+    setConfirmDialogOpen(false);
+  }, [setConfirmDialogOpen]);
```
Comment (ajohn25) [inline src/containers/AdminCampaignEdit/components/StartCampaignButton.tsx:63]: Nit: I think this doesn't need a dependency or maybe even fine to render this without `useCallback()` ?
```suggestion
  const handleCloseDialog = useCallback(() => {
    setConfirmDialogOpen(false);
  }, []);
```

File: src/containers/AdminCampaignEdit/components/StartCampaignButton.tsx [hunk 6]
```diff
@@ -38,11 +47,36 @@ export const StartCampaignButton: React.FC<StartCampaignButtonProps> = (
...
+    }
+
     startCampaign({ variables: { campaignId } });
-  }, [startCampaign, campaignId]);
+  }, [
+    startCampaign,
+    campaignId,
+    campaignHasSurveyResponses,
+    warningConfirmed,
+    setConfirmDialogOpen
```
Comment (ajohn25) [inline src/containers/AdminCampaignEdit/components/StartCampaignButton.tsx:78]: Nit:
```suggestion
    warningConfirmed
```

File: src/containers/AdminCampaignEdit/components/StartCampaignButton.tsx [hunk 7]
```diff
@@ -38,11 +47,36 @@ export const StartCampaignButton: React.FC<StartCampaignButtonProps> = (
...
+  const campaignInteractionSteps =
+    campaignStepsData?.campaign?.interactionSteps ?? [];
+
+  const campaignHasSurveyResponses = campaignInteractionSteps.length > 1;
+
+  const handleConfirmWarningAndStartCampaign = useCallback(() => {
+    setWarningConfirmed(true);
+    setConfirmDialogOpen(false);
+    startCampaign({ variables: { campaignId } });
+  }, [startCampaign, campaignId, setWarningConfirmed, setConfirmDialogOpen]);
```
Comment (ajohn25) [inline src/containers/AdminCampaignEdit/components/StartCampaignButton.tsx:59]: Nit:
```suggestion
  }, [startCampaign, campaignId]);
```

**Review by ajohn25 — APPROVED** [review]
Sorry I noticed a new set of Nits 😭 All comments based on this piece of info below. I'm not too familiar with `useCallback()` , just did a quick search, so also feel free to judge for yourself what's best!

> React guarantees that setState function identity is stable and won’t change on re-renders. This is why it’s safe to omit from the useEffect or useCallback dependency list.

https://legacy.reactjs.org/docs/hooks-reference.html#usecallback

**Comment by lediur** [comment]
What do we think about rearranging the body text of the modal to move the "For the best experience..." to the beginning, then adding a paragraph break after the sentence?

I think it's quite a bit of text, but rearranging could improve readability.

So:

```
For the best experience...

Tags do not prepopulate scripts... (Read more in docs [external link icon])
```

i could take or leave the external link icon

**Review by ajohn25 — APPROVED** [review]

**Pull request merged**